src/System.Private.CoreLib/src/System/ValueType.cs

```diff
@@ -36,4 +42,161 @@ public override int GetHashCode()
+                for (int i = 0; i < valueTypeSize; i++)
+                {
+                    if (Unsafe.Add(ref thisRawData, i) != Unsafe.Add(ref thatRawData, i))
+                        return false;
+                }
+            }
+            else
+            {
+                // Foreach field, box and call the Equals method.
+                for (int i = 0; i < numFields; i++)
+                {
+                    int fieldOffset = __GetFieldHelper(i, out EETypePtr fieldType);
+
+                    // Fetch the value of the field on both types
+                    object thisField = RuntimeImports.RhBoxAny(ref Unsafe.Add(ref thisRawData, fieldOffset), fieldType);
+                    object thatField = RuntimeImports.RhBoxAny(ref Unsafe.Add(ref thatRawData, fieldOffset), fieldType);
+
+                    // Compare the fields
+                    if (thisField == null)
+                    {
+                        if (thatField != null)
+                            return false;
+                    }
+                    else if (!thisField.Equals(thatField))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        public override int GetHashCode()
+        {
+            int hashCode = this.EETypePtr.GetHashCode();
+
+            int numFields = __GetFieldHelper(GetNumFields, out _);
+
+            if (numFields == UseFastHelper)
+            {
+                hashCode ^= FastGetValueTypeHashCodeHelper(this.EETypePtr, ref this.GetRawData());
+            }
+            else
+            {
+                hashCode ^= RegularGetValueTypeHashCode(this.EETypePtr, ref this.GetRawData(), numFields);
+            }
+
+            return hashCode;
+        }
+
+        private static int FastGetValueTypeHashCodeHelper(EETypePtr type, ref byte data)
+        {
+            // Sanity check - if there are GC references, we should not be hashing bytes
+            Debug.Assert(!type.HasPointers);
+
+            int size = (int)type.ValueTypeSize;
+            int hashCode = 0;
+
+            for (int i = 0; i < size / 4; i++)
+            {
+                hashCode ^= Unsafe.As<byte, int>(ref Unsafe.Add(ref data, i * 4));
+            }
+
+            return hashCode;
+        }
+
+        private int RegularGetValueTypeHashCode(EETypePtr type, ref byte data, int numFields)
+        {
+            int hashCode = 0;
+
+            // We only take the hashcode for the first non-null field. That's what the CLR does.
+            for (int i = 0; i < numFields; i++)
+            {
+                int fieldOffset = __GetFieldHelper(i, out EETypePtr fieldType);
+                ref byte fieldData = ref Unsafe.Add(ref data, fieldOffset);
+
+                Debug.Assert(!fieldType.IsPointer);
+
+                if (fieldType.CorElementType == RuntimeImports.RhCorElementType.ELEMENT_TYPE_R4)
```

Comment:
While this looks like a good idea, CoreCLR just pushes this down the XOR path I complained about below

Reply:
This was supposed to be fixed by https://github.com/dotnet/coreclr/pull/13164. Opened: https://github.com/dotnet/coreclr/issues/16545